writer_test.go

```diff
@@ -49,3 +52,182 @@ func TestWriter_multipartWithoutBoundary(t *testing.T) {
+
+Unfortunately, we cannot rely on the go defer statement to have
+RemoveTab(), called *AFTER* c.Unlock(). Deferred statements are called
+in the reverse order they were evaluated.
+
+Use an explicit function array to ensure that the composer is unlocked
+before RemoveTab() is called.
+
+Fixes: 6b97085ae5ec ("tab: fix broken history on removal")
+Signed-off-by: Robin Jarry <robin@jarry.cc>
+---
+ app/compose.go | 24 ++++++++++++++++++------
+ 1 file changed, 18 insertions(+), 6 deletions(-)
+
+diff --git a/app/compose.go b/app/compose.go
+index 7a6a423c3ea3..e8a672479245 100644
+--- a/app/compose.go
++++ b/app/compose.go
+@@ -1197,7 +1197,15 @@ func (c *Composer) reopenEmailFile() error {
+ 
+ func (c *Composer) termClosed(err error) {
+ 	c.Lock()
+-	defer c.Unlock()
++	// RemoveTab() on error must be called *AFTER* c.Unlock() but the defer
++	// statement does the exact opposite (last defer statement is executed
++	// first). Use an explicit list that begins with unlocking first.
++	deferred := []func(){c.Unlock}
++	defer func() {
++		for _, d := range deferred {
++			d()
++		}
++	}()
+ 	if c.editor == nil {
+ 		return
+ 	}
+@@ -1205,7 +1213,7 @@ func (c *Composer) termClosed(err error) {
+ 		PushError("Failed to reopen email file: " + e.Error())
+ 	}
+ 	editor := c.editor
+-	defer editor.Destroy()
++	deferred = append(deferred, editor.Destroy)
+ 	c.editor = nil
+ 	c.focusable = c.focusable[:len(c.focusable)-1]
+ 	if c.focused >= len(c.focusable) {
+@@ -1213,8 +1221,10 @@ func (c *Composer) termClosed(err error) {
+ 	}
+ 
+ 	if editor.cmd.ProcessState.ExitCode() > 0 {
+-		RemoveTab(c, true)
+-		PushError("Editor exited with error. Compose aborted!")
++		deferred = append(deferred, func() {
++			RemoveTab(c, true)
++			PushError("Editor exited with error. Compose aborted!")
++		})
+ 		return
+ 	}
+ 
+@@ -1225,8 +1235,10 @@ func (c *Composer) termClosed(err error) {
+ 			PushError(err.Error())
+ 			err := c.showTerminal()
+ 			if err != nil {
+-				RemoveTab(c, true)
+-				PushError(err.Error())
++				deferred = append(deferred, func() {
++					RemoveTab(c, true)
++					PushError(err.Error())
++				})
+ 			}
+ 			return
+ 		}
+-- 
+2.47.1
+
+
+`, "\n", "\r\n")
+
+	msg, err := Read(strings.NewReader(original))
+	if err != nil {
+		t.Fatalf("Read: %v", err)
+	}
+
+	msg.Header.Del("Sender")
+
+	var buf bytes.Buffer
+	err = msg.WriteTo(&buf)
+	if err != nil {
+		t.Fatalf("WriteTo: %v", err)
+	}
+
+	if buf.String() != original {
+		orig, _ := os.CreateTemp(os.TempDir(), "go-message-orig-*.txt")
+		after, _ := os.CreateTemp(os.TempDir(), "go-message-after-*.txt")
+		defer os.Remove(orig.Name())
+		defer os.Remove(after.Name())
+		orig.WriteString(original)
+		orig.Close()
+		after.Write(buf.Bytes())
+		after.Close()
+
+		var stdout bytes.Buffer
+		cmd := exec.Command("diff", "-u", orig.Name(), after.Name())
+		cmd.Stdout = &stdout
+		cmd.Run()
+		t.Fatalf("reformatted message differs:\n%s", stdout.String())
```

Comment:
Hm, not a fan of shelling out to `diff`. Even if that makes reading the difference more difficult, can we just print both strings instead?

Reply:
yes, it was to make spotting the differences easier, but I guess you can instrument the test later to debug it more.